Lab_1/Aula_rest/src/application.c

```diff
@@ -9,28 +20,141 @@
  * @param x valor de x em "/dev/ttySx"
  * @param type RECEIVER|TRANSMITTER
  */
-void parseArgs(int argc, char** argv, int* x, int* type){
-  if (argc != 3){
-    log_error("Usage: ./application [receiver/transmitter] [nPorta = {0,1,10,11}]\n");
+void parseArgs(int argc, char** argv){
+  if (argc != 4){
+    log_error("Usage: ./application (receiver <destination> | transmitter <filename>) [gate = {0,1,10,11}]\n");
     exit(-1);
   }
 
-  *x = atoi(argv[2]);
-  if (*x != 0 && *x != 1 && *x != 10 && *x != 11) {
-    log_caution("nPorta = {0,1,10,11}\n");
-    exit(-1);
+  applayer.gate = atoi(argv[3]);
+  if (applayer.gate != 0 && applayer.gate != 1 && applayer.gate != 10 && applayer.gate != 11) {
+      log_caution("nPorta = {0,1,10,11}\n");
+      exit(-1);
   }
 
   if (strcmp("receiver",argv[1])==0){
-    *type = RECEIVER;
-  } else if (strcmp("transmitter",argv[1])==0){
-    *type = TRANSMITTER;
-  } else {
-    log_caution("Usage: ./application [receiver/transmitter] [nPorta = {0,1,10,11}]\n");
-    exit(-1);
+      applayer.type = RECEIVER;
+      strcpy(applayer.destinationArg,argv[2]);
+  } 
+  else if (strcmp("transmitter",argv[1])==0){
+      applayer.type = TRANSMITTER;
+      strcpy(applayer.filenameArg,argv[2]);
+  } 
+  else {
+      log_caution("Usage: ./application (receiver <destination> | transmitter <filename>) [gate = {0,1,10,11}]\n");
+      exit(-1);
   }
 }
 
+int transmitterApp(int fd, char *file){
+  char controlPackage[255];
+  char dataPackage[MAX_SIZE];
+  int sequenceNumber = 0;
+  int nbytes;
+  int file_fd;
+  char * file_data[MAX_SIZE];
+
+  struct stat fileInfo;
+
+  if (stat(file, &fileInfo)<0){
+    log_error("Error obtaining file information.\n");
+    return -1;
+  }
+
+  file_fd = open(file,O_RDONLY);
+  if (file_fd <0){
+    log_error("Error opening file.\n");
+    return -1;
+  }
+
+  /*building control package*/
+  controlPackage[0] = START;
+  controlPackage[1] = T_SIZE;
+  controlPackage[2] = sizeof(fileInfo.st_size); /* sending size of off_t */
+  memcpy(&controlPackage[3],&fileInfo.st_size,sizeof(fileInfo.st_size));
+
+  controlPackage[sizeof(fileInfo.st_size)+3] = T_NAME;
+  controlPackage[sizeof(fileInfo.st_size)+4] = strlen(file);
+  memcpy(&controlPackage[sizeof(fileInfo.st_size)+5],file,strlen(file));
+
+  llwrite(fd,controlPackage,sizeof(fileInfo.st_size) + 5 + strlen(file));
+
+  while( (nbytes = read(file_fd,file_data,MAX_SIZE-4)) != 0){
+    /*building data package*/
+    dataPackage[0] = DATA;
+    dataPackage[1] = sequenceNumber % 255;
+    dataPackage[2] = nbytes / 256;
+    dataPackage[3] = nbytes % 256;
+    memcpy(&dataPackage[4],file_data,nbytes);
+
+    llwrite(fd,dataPackage,nbytes+4);
+
+    sequenceNumber++;
+  }
+
+  controlPackage[0] = END;
+
+  llwrite(fd,controlPackage,sizeof(fileInfo.st_size) + 5 + strlen(file));
+  return 0;
+}
+
+void parseFileInfo(unsigned char *controlpackage){
```

Comment:
A forma de assegurar que os bytes `T` aparecem na ordem `T_SIZE` -> `T_NAME` é feita na nossa escrita mas se for usado outro programa de escrita esta ordem não é assegurada. Se calhar ficaria aqui melhor um ciclo que vai lendo byte a byte e processando de acordo com o que encontra

Reply:
Yap concordo